Mathlib/RingTheory/AlgebraicIndependent/TranscendenceBasis.lean

```diff
@@ -125,3 +208,299 @@ theorem IsTranscendenceBasis.isAlgebraic_field {F E : Type*} {x : ι → E}
   haveI : IsScalarTower (adjoin F S) (IntermediateField.adjoin F S) E :=
     IsScalarTower.of_algebraMap_eq (congrFun rfl)
   exact Algebra.IsAlgebraic.extendScalars (R := adjoin F S) (Subalgebra.inclusion_injective _)
+
+namespace AlgebraicIndependent
+
+variable (inj : Injective (algebraMap R A))
+
+section
+
+variable [NoZeroDivisors A]
+
+private def indepMatroid : IndepMatroid A where
+  E := univ
+  Indep := Set.AlgebraicIndependent R
+  indep_empty := (algebraicIndependent_empty_iff ..).mpr inj
+  indep_subset _ _ := (·.mono)
+  indep_aug I B I_ind h B_base := by
+    contrapose! h
+    rw [← isTranscendenceBasis_iff_maximal] at B_base ⊢
+    cases subsingleton_or_nontrivial R
+    · rw [isTranscendenceBasis_iff_of_subsingleton] at B_base ⊢
+      contrapose! h
+      have ⟨b, hb⟩ := B_base
+      exact ⟨b, ⟨hb, fun hbI ↦ h ⟨b, hbI⟩⟩, .of_subsingleton⟩
+    rw [I_ind.isTranscendenceBasis_iff_isAlgebraic]
+    replace B_base := B_base.isAlgebraic
+    rw [Subtype.range_val] at B_base ⊢
+    refine ⟨fun a ↦ (B_base.1 a).adjoin_of_forall_isAlgebraic fun x hx ↦ ?_⟩
+    contrapose! h
+    exact ⟨x, hx, I_ind.insert h⟩
+  indep_maximal X _ I ind hIX := exists_maximal_algebraicIndependent I X hIX ind
+  subset_ground _ _ := subset_univ _
+
+/-- If `R` is a commutative ring and `A` is a commutative `R`-algebra with injective algebra map
+and no zero-divisors, then the `R`-algebraic independent subsets of `A` form a matroid. -/
+def matroid : Matroid A := (indepMatroid inj).matroid.copyBase univ
+  (fun s ↦ IsTranscendenceBasis R ((↑) : s → A)) rfl
+  (fun B ↦ by simp_rw [Matroid.base_iff_maximal_indep, isTranscendenceBasis_iff_maximal]; rfl)
+
+instance : (matroid inj).Finitary where
+  indep_of_forall_finite := algebraicIndependent_of_finite
+
+@[simp] theorem matroid_e : (matroid inj).E = univ := rfl
+
+theorem matroid_indep_iff {s : Set A} :
+    (matroid inj).Indep s ↔ s.AlgebraicIndependent R := Iff.rfl
+
+theorem matroid_base_iff {s : Set A} :
+    (matroid inj).Base s ↔ IsTranscendenceBasis R ((↑) : s → A) := Iff.rfl
+
+theorem matroid_cRank_eq : (matroid inj).cRank = trdeg R A :=
+  (trdeg_eq_iSup_cardinalMk_isTranscendenceBasis _).symm
+
+end
+
+theorem matroid_basis_iff [IsDomain A] {s t : Set A} : (matroid inj).Basis s t ↔
```

Comment:
@apnelson1 
I've established algebraic characterization of `Basis`, `closure`, `Flat`, and `Spanning` for the algebraic independence matroid starting from here. It turns out it's easier to first characterize `closure` before proceeding to `Flat`. I also added two missing Matroid lemmas which I'll add to #21512.